lib/puma/thread_pool.rb

```diff
@@ -228,6 +228,25 @@ def wait_until_not_full
       end
     end
 
+    def wait_for_less_busy_worker(delay_s)
+      # Ruby MRI does GVL, this can result
+      # in processing contention when multiple threads
+      # (requests) are running concurrently
+      return unless Puma.mri?
```

Comment:
This is adding two method calls to the hot path, but for now I'll leave them in. Just pointing out for future optimiation that we could metaprogram these out based on config.

Reply:
It might be worth swapping the order of calls.

```
require "benchmark/ips"

def mri?
  RUBY_ENGINE == 'ruby' || RUBY_ENGINE.nil?
end

num = rand(-10..10)

Benchmark.ips do |b|
  b.report("mri?") { mri? }

  b.report("delay_s > 0") { num > 0 }

  b.compare!
end
```

```
Warming up --------------------------------------
                mri?   714.254k i/100ms
         delay_s > 0     1.675M i/100ms
Calculating -------------------------------------
                mri?      7.118M (± 3.3%) i/s -     35.713M in   5.023186s
         delay_s > 0     15.494M (±10.6%) i/s -     77.054M in   5.055471s

Comparison:
         delay_s > 0: 15493847.0 i/s
                mri?:  7118414.2 i/s - 2.18x  (± 0.00) slower
```

(I'm not sure how hot this path runs / how much it matters.)

Reply:
https://github.com/puma/puma/pull/2579

Reply:
Would ruby 2.2.1 be impacted by MRI GVL as well?